KEP 2535: Ensure Secret Pulled Images (#49886)
kubernetes/website: content/en/docs/reference/node/kubelet-files.md

[Graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
stores state locally at `/var/lib/kubelet/graceful_node_shutdown_state`.

### Image Pull Records

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

The kubelet stores a record of attempted and successful image pulls, and uses it to

> bypass registry authentication when an image is re-pulled with the same credentials.

Reviewer: This is not very accurate. There is no bypass happening since the image is already on the node. The credentials cannot be directly used because their plaintext is never stored.

These records are cached and also stored as files in the kubelet's base directory,

> under a subdirectory `image_manager`. This subdirectory has two subdirectories:

Reviewer: Mention the full path as in the examples above.

It might make sense to have the two subdirectories turned into list items and explain the files that get stored in each.

[Graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
stores state locally at `/var/lib/kubelet/graceful_node_shutdown_state`.

### Image Pull Records

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

The kubelet stores records of attempted and successful image pulls, and uses it to
verify that the image was previously successfully pulled with the same credentials.

These records are cached as files in the `image_registry` directory within the 

> kubelet base directory, so on a typical Linux node this means `/var/lib/kubelet/image_manager`.

Reviewer 2: ```suggestion
kubelet base directory. On a typical Linux node, this means `/var/lib/kubelet/image_manager`.
```

There are two subdirectories to `image_manager`:
* `pulling` - stores records about images the Kubelet is attempting to pull.

> * `pulled` - stores records about images that were successfully pulled by the Kubelet, along with
> metadata about the credentials used for the pulls.

Reviewer 2: ```suggestion
* `pulled` - stores records about images that were successfully pulled by the Kubelet, 
  along with metadata about the credentials used for the pulls.
```

> See [private registries](/docs/concepts/containers/images#ensureimagepullcredentialverification)

Reviewer: ```suggestion
See [Ensure Image Pull Credential Verification](/docs/concepts/containers/images#ensureimagepullcredentialverification)
```

[Graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
stores state locally at `/var/lib/kubelet/graceful_node_shutdown_state`.

### Image Pull Records

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

The kubelet stores records of attempted and successful image pulls, and uses it to
verify that the image was previously successfully pulled with the same credentials.

> These records are cached and also stored as files in the `plugins_registry` within the 

Reviewer: > These records are cached and also stored

the files are the cache, currently

```suggestion
These records are cached as files in the `image_registry` directory within the 
```

kubelet base directory, so on a typical Linux node this means `/var/lib/kubelet/image_manager`.

> This subdirectory has two subdirectories:

Reviewer: ```suggestion
There are two subdirectories to `image_manager`:
```

* `pulling` - stores records about images the Kubelet is attempting to pull.

> * `pulled` - stores records about images that were successfully pulled by the Kubelet, along with
> metadata about the credentials used, the secret reference and a SHA-256 hash of the secret used 
> to authenticate the image pull.

Reviewer: ```suggestion
* `pulled` - stores records about images that were successfully pulled by the Kubelet, along with
metadata about the credentials used for the pulls.
```

[Graceful node shutdown](/docs/concepts/cluster-administration/node-shutdown/#graceful-node-shutdown)
stores state locally at `/var/lib/kubelet/graceful_node_shutdown_state`.

### Image Pull Records

{{< feature-state feature_gate_name="KubeletEnsureSecretPulledImages" >}}

The kubelet stores a record of attempted and successful image pulls, and uses it to

> verify that the image was previously authenticated with the same credentials. The 

Reviewer: ```suggestion
verify that the image was previously successfully pulled with the same credentials. The 
```

image digest, secret reference, and a hash of the secret used are stored to be verified

> against future image pulls.

Reviewer: > The image digest, secret reference, and a hash of the secret used are stored to be verified
against future image pulls.

drop this part

These records are cached and also stored as files in the kubelet's base directory,

> under `/var/lib/kubelet/image_manager`. This subdirectory has two subdirectories:

Reviewer: Let's stick to the longer description of the dir, like in the DRA case, citing:
```
The directory that the kubelet looks in is `plugins_registry` within the kubelet base
directory, so on a typical Linux node this means `/var/lib/kubelet/plugins_registry`.
```

> * `pulling`, which stores image digests that the kubelet is intending to verify and pull.

Reviewer: ```suggestion
* `pulling` - stores records about images the Kubelet is attempting to pull.
```

> * `pulled`, which stores the digest and name of any successfully pulled image along

Reviewer: something like this maybe
```suggestion
* `pulled` - stores records about images that were successfully pulled by the Kubelet, along with metadata about the credentials used for the pull 
```
